src/wasm/asm-wasm-builder.cc

```diff
@@ -317,27 +362,48 @@ class AsmWasmBuilderImpl : public AstVisitor {
     RECURSE(Visit(expr->expression()));
   }
 
+#define ADD_ALL_TYPES(token, op)                    \
+  m[std::make_pair(token, kAstI32)] = kExprI32##op; \
+  m[std::make_pair(token, kAstF32)] = kExprF32##op; \
+  m[std::make_pair(token, kAstF64)] = kExprF64##op;
+
+  typedef std::pair<Token::Value, LocalType> op_map_key;
+
+  static std::map<op_map_key, WasmOpcode> binary_op_map() {
```

Comment:
This looks pretty expensive. It will instantiate and build a tree map every time it's requested, and copy it. How about a flat static array indexed by LocalType?